### ISSUE TITLE
The parser does not like long boolean expressions

### ISSUE THREAD

**Issue by MDoerner** [rubberduck-vba/Rubberduck#3634]
The problem is that the parser slows down to a crawl in the presence of long boolean expressions. The runtime seems to be roughly exponential in the number of concatenated expressions.

Examples:

From #3556 
```vba
If preserveUserColor = True _
   And rCell.row > 6 _
   And Not cColor = matColor("blue", 50) _
   And Not cColor = matColor("blue", 100) _
   And Not cColor = matColor("blue", 200) _
   And Not cColor = matColor("blue", 300) _
   And Not cColor = matColor("blue", 400) _
   And Not cColor = matColor("blue", 500) _
   And Not cColor = matColor("blue", 600) _
   And Not cColor = matColor("blue", 700) _
   And Not cColor = matColor("blue", 800) _
   And Not cColor = matColor("blue", 900) _
   And Not cColor = matColor("amber", 100) _
   And Not cColor = matColor("amber", 200) _
   And Not cColor = matColor("lime", 100) _
   And Not cColor = matColor("lime", 200) _
   And Not cColor = matColor("bluegrey", 100) _
   And Not cColor = matColor("bluegrey", 200) _
   And Not cColor = matColor("deeppurple", 100) _
   And Not cColor = matColor("deeppurple", 200) _
   And Not cColor = matColor("yellow", 200) _
   And Not cColor = matColor("teal", 200) _
Then GoTo skiprCellFor
```

From chat:

```vba
Dim Ignore As Boolean
Ignore = TypeOf Me Is Class1 And TypeOf Me Is Class1 And _
TypeOf Me Is Class1 And TypeOf Me Is Class1 And _
TypeOf Me Is Class1 And TypeOf Me Is Class1 And _
TypeOf Me Is Class1 And TypeOf Me Is Class1 And _
TypeOf Me Is Class1 And TypeOf Me Is Class1 And _
TypeOf Me Is Class1 And TypeOf Me Is Class1 And _
TypeOf Me Is Class1 And TypeOf Me Is Class1 And _
TypeOf Me Is Class1 
```

My guess is that the parser is constantly guessing the wrong way to proceed. Reordering the subrules of expression might help. Upgrading Antlr might help as well, since the left-recursion resollution changed. (This would make changes to the ordering necesary anyway in order to get the SLL parser to work again.)

ref #2498

**Comment by parrt** [comment]
What is the target language? C#?

**Comment by Hosch250** [comment]
Yes, we are a C# project. We are parsing VBA code.

**Comment by retailcoder** [comment]
@parrt correct. Are we doomed? Our grammar is here if you want to have a look - it's based on the VBA language specifications: https://github.com/rubberduck-vba/Rubberduck/blob/next/Rubberduck.Parsing/Grammar/VBAParser.g4

**Comment by Hosch250** [comment]
Also, feel free to visit our warroom here: https://chat.stackexchange.com/rooms/14929/vba-rubberducking

**Comment by parrt** [comment]
Not doomed but might have to convert https://github.com/rubberduck-vba/Rubberduck/blob/next/Rubberduck.Parsing/Grammar/VBAParser.g4#L588 to avoid left-recursion. Are you using latest antlr 4.7.1?

**Comment by retailcoder** [comment]
@parrt still on v4.3... seems there have been a number of breaking changes we're struggling with. Conversion is in progress though. Not sure avoiding recursion is possible with that one (not to mention all the implications in our resolver code).. hopefully 4.7 works better with it?

**Comment by parrt** [comment]
There have been important optimizations in later version so definitely upgrade

**Comment by ThunderFrame** [comment]
Some testing...

- I'm getting odd behavior with refresh times - it's almost like ANTLR is caching a parsing strategy... The first parse takes some time, but with small edits, the subsequent parse is very fast.

- I'm not sure that it's the multiple boolean conditions per se, but rather the complexity of the expressions

This parses/resolves in **4 seconds**:
```vb
Sub test()
  Debug.Print False And _
              False And _
              False And _
              False And _
              False And _
              False And _
              False And _
              False And _
              False And _
              False And _
              False And _
              False And _
              False
End Sub
```

Whereas, this resolves in **12 seconds**:
```vb
Sub test()
  Debug.Print Not False And _
              Not False And _
              Not False And _
              Not False And _
              Not False And _
              Not False And _
              Not False And _
              Not False And _
              Not False And _
              Not False And _
              Not False And _
              Not False And _
              Not False
End Sub
```

But, if I edit that code to substitute `False` for `True`, and reparse, and then revert `True` to `False` and reparse, it resolves in 4 seconds. It's like ANTLR has cached the parsing strategy for this statement. If I add an *extra* condition and reparse, the parse time shoots up again, and then if I remove the extra condition and reparse, the parse time is low again, which suggests that the cached optimizations are independent of individual reparses?

And this is **still resolving after 10 minutes**...
```vb
Sub test()
  Debug.Print Not 42 = 42 And _
              Not 42 = 42 And _
              Not 42 = 42 And _
              Not 42 = 42 And _
              Not 42 = 42 And _
              Not 42 = 42 And _
              Not 42 = 42 And _
              Not 42 = 42 And _
              Not 42 = 42 And _
              Not 42 = 42 And _
              Not 42 = 42 And _
              Not 42 = 42 And _
              Not 42 = 42
End Sub
```

**Comment by parrt** [comment]
Definitely caches partial parsing results so that similar parses will get the benefit of reusing the analysis.  It warms up just like a jit.

**Comment by parrt** [comment]
Are you using two-stage parsing? SLL *then* full ALL only if SLL fails?

**Comment by retailcoder** [comment]
@parrt Yup. We attempt a quick parse first, and fallback to the slower prediction mode if the initial pass throws a parse exception.

**Comment by parrt** [comment]
okidoki. we had a big optimization for left-recursive expressions in a recent release, so definitely upgrade.

**Comment by Vogel612** [comment]
yes. We first attempt to parse with SLL to get the performance boost. In something like <= 5% of cases (from logfiles posted in issues) SLL will fail, and we retry parsing the whole module with LL prediction. See [VBAModuleParser](https://github.com/rubberduck-vba/Rubberduck/blob/next/Rubberduck.Parsing/VBA/VBAModuleParser.cs) for code

**Comment by Vogel612** [comment]
note that we're running multiple parse passes that way, because Attributes are only available when exporting a module. We need the line numbers to match with the displayed code though, so there's two runs for each module. 

Additionally we handle compilation directives before each of these runs.

**Comment by MDoerner** [comment]
As @Vogel612 already showed above, the complexity of the boolean expression seems to be an important factor. I think the basic problem here is that the parser has a hard time sorting out where the operands of `And` and `Or` begin and end. In the presence of parentheses, the parse is much faster.

Parses in about 1s:
```vba
Public Sub Test()
    Dim Ignore As Boolean
    Ignore = TypeOf Me Is Class1 And TypeOf Me Is Class1 And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1) And _
    (TypeOf Me Is Class1) And (TypeOf Me Is Class1)
End Sub
```

Parses in about 380s:
```vba
Public Sub Test()
    Dim Ignore As Boolean
    Ignore = TypeOf Me Is Class1 And TypeOf Me Is Class1 And _
    TypeOf Me Is Class1 And TypeOf Me Is Class1 And _
    TypeOf Me Is Class1 And TypeOf Me Is Class1 And _
    TypeOf Me Is Class1 And TypeOf Me Is Class1 And _
    TypeOf Me Is Class1 And TypeOf Me Is Class1 And _
    TypeOf Me Is Class1 And TypeOf Me Is Class1
End Sub
```

**Comment by MDoerner** [comment]
I can confirm that this has been fixed by the move to Antlr 4.6.4 (PR #3715).

Now the parse for the examples above is basically instant.
